Source/Basetypes/BigFloat.cs

```diff
@@ -523,73 +434,158 @@ public void FloorCeiling(out BIM floor, out BIM ceiling)
+      }
 
-    public bool IsSpecialType {
-      get {
-        if (value == "")
-          return false;
-        return (value.Equals("NaN") || value.Equals("+oo") || value.Equals("-oo"));
+      BIM xsig = x.significand, ysig = y.significand;
+      BIM xexp = x.exponent, yexp = y.exponent;
+
+      BIM hiddenBitPow = BIM.Pow(2, x.significandSize - 1);
+
+      if (xexp > 0) {
+        xsig += hiddenBitPow;
+      } else {
+        ++xexp;
       }
-    }
 
-    public bool IsPositive {
-      get {
-        return !IsNegative;
+      if (yexp > 0) {
+        ysig += hiddenBitPow;
+      } else {
+        ++yexp;
       }
+
+      ysig *= xsig;
+      yexp += xexp - (BIM.Pow(2, x.exponentSize - 1) - 1) - (x.significandSize - 1);
+
+      while (ysig >= hiddenBitPow * 2 || yexp <= 0) {
+        ysig >>= 1;
+        ++yexp;
+      }
+
+      while (ysig < hiddenBitPow && yexp > 1) {
+        ysig <<= 1;
+        --yexp;
+      }
+
+      if (ysig < hiddenBitPow) {
+        yexp = 0;
+      } else {
+        ysig -= hiddenBitPow;
+      }
+
+      if (yexp >= BIM.Pow(2, x.exponentSize) - 1) {
+        return new BigFloat(x.isSignBitSet ^ y.isSignBitSet ? "-oo" : "+oo", x.significandSize, x.exponentSize);
+      }
+
+      return new BigFloat(x.isSignBitSet ^ y.isSignBitSet, ysig, yexp, x.significandSize, x.exponentSize);
     }
 
+
+    ////////////////////////////////////////////////////////////////////////////
+    // Some basic comparison operations
+
     public bool IsZero {
       get {
-        return significand.Equals(BigNum.ZERO) && Exponent == BIM.Zero;
+        return value == "" && significand.IsZero && exponent.IsZero;
       }
     }
 
     [Pure]
     public int CompareTo(BigFloat that) {
-      if (this.exponent > that.exponent)
-        return 1;
-      if (this.exponent < that.exponent)
-        return -1;
-      if (this.significand == that.significand)
+      Contract.Requires(exponentSize == that.exponentSize);
+      Contract.Requires(significandSize == that.significandSize);
+
+      if (value == "" && that.value == "") {
+        int cmpThis = IsZero ? 0 : isSignBitSet ? -1 : 1;
+        int cmpThat = that.IsZero ? 0 : that.isSignBitSet ? -1 : 1;
+
+        if (cmpThis == cmpThat) {
+          if (exponent == that.exponent) {
+            return cmpThis * significand.CompareTo(that.significand);
+          }
+
+          return cmpThis * exponent.CompareTo(that.exponent);
+        }
+
+        if (cmpThis == 0) {
+          return -cmpThat;
+        }
+
+        return cmpThis;
+      }
+
+      if (value == that.value) {
         return 0;
-      return this.significand > that.significand ? 1 : -1;
+      }
+
+      if (value == "NaN" || that.value == "+oo" || value == "-oo" && that.value != "NaN") {
+        return -1;
+      }
+
+      return 1;
     }
 
     [Pure]
     public static bool operator ==(BigFloat x, BigFloat y) {
+      if (x.value == "NaN" || y.value == "NaN") {
```

Comment:
Why do you need these checks for `NaN` here?
To me it seems that your intention was to handle special cases in the `compareTo` method as well.

Reply:
`CompareTo` isn't supposed to return results that are consistent with what the `==, !=, >, <, >=, <=` operators return. Specifically, if any of the 6 operators are given a `NaN` value for at least one of their operands, they should always return false. However, I think `BigFloat.CompareTo` should follow the specification for `Single.CompareTo`, described here:

https://docs.microsoft.com/en-us/dotnet/api/system.single.compareto?view=netframework-4.8

As described in that document, `NaNs` are treated differently than for the 6 operators. For example, the expression `(0.0f / 0.0f).CompareTo(0.0f / 0.0f)` should return true, whereas the expression `(0.0f / 0.0f) == (0.0f / 0.0f)` should return false.

Reply:
Except `!=`, which should always return true if `NaN` is passed.

Reply:
Ah, great, I get it, thanks for clarifying this. It would be useful to add what you just wrote here (a shorter summary of it) as a comment on the `CompareTo` method.

Reply:
Alright, I've added this.